File: frontend/src/engine/loader.ts
// CSV loading + validation. Mirrors backend/app/data/loader.py.

import type { Bar, Dataset } from "./types";

const REQUIRED = ["time", "open", "high", "low", "close"];

export function parseCsv(text: string, filename?: string): Dataset {
  const lines = text.split(/\r?\n/).filter((l) => l.trim().length > 0);
  if (lines.length < 2) throw new Error("CSV has no data rows.");

  const header = splitRow(lines[0]).map((h) => h.trim().toLowerCase());
  const idx: Record<string, number> = {};
  for (const col of [...REQUIRED, "volume"]) {
    const i = header.indexOf(col);
    if (i >= 0) idx[col] = i;
  }
  const missing = REQUIRED.filter((c) => !(c in idx));
  if (missing.length) {
    throw new Error(
      `CSV is missing required column(s): ${missing.join(", ")}. Found: ${header.join(", ")}`
    );
  }

  const firstTime = splitRow(lines[1])[idx.time];
  const sourceOffset = extractOffset(firstTime);

  const bars: Bar[] = [];
  let precision = 0;
  for (let r = 1; r < lines.length; r++) {
    const cells = splitRow(lines[r]);
    const rawTime = cells[idx.time];
    const ms = new Date(rawTime).getTime();
    if (Number.isNaN(ms)) throw new Error(`Could not parse timestamp: "${rawTime}"`);

    const open = Number(cells[idx.open]);
    const high = Number(cells[idx.high]);
    const low = Number(cells[idx.low]);
    const close = Number(cells[idx.close]);
    if ([open, high, low, close].some((v) => Number.isNaN(v))) continue;
    const volume = "volume" in idx ? Number(cells[idx.volume]) || 0 : 0;

    // Track the most decimal places seen, to size P/L formatting per instrument.
    for (const col of ["open", "high", "low", "close"] as const) {
      precision = Math.max(precision, decimalsOf(cells[idx[col]]));
    }

    bars.push({ ms, open, high, low, close, volume });
  }

  if (bars.length === 0) throw new Error("No valid rows remained after parsing.");

  // Sort ascending and drop duplicate timestamps (keep last).
  bars.sort((a, b) => a.ms - b.ms);
  const deduped: Bar[] = [];
  for (const b of bars) {
    const last = deduped[deduped.length - 1];
    if (last && last.ms === b.ms) deduped[deduped.length - 1] = b;
    else deduped.push(b);
  }

  return {
    bars: deduped,
    instrument: inferInstrument(filename),
    interval_minutes: inferInterval(deduped),
    source_offset: sourceOffset,
    price_precision: precision,
  };
}

// Decimal places in a numeric string like "1.08345" (-> 5). Capped at 8.
function decimalsOf(s: string): number {
  if (!s) return 0;
  const dot = s.indexOf(".");
  if (dot < 0) return 0;
  return Math.min(8, s.trim().length - dot - 1);
}

function splitRow(line: string): string[] {
  // OANDA/TradingView exports are simple comma-separated with no quoted commas.
  return line.split(",");
}

function inferInstrument(filename?: string): string {
  if (!filename) return "UNKNOWN";
  const stem = filename.split("/").pop()!.replace(/\.[^.]+$/, "");
  const parts = stem.split("_");
  for (let i = 0; i < parts.length; i++) {
    // The OANDA token may be hyphen-prefixed, e.g. "5f7f2016-OANDA".
    if (parts[i].toUpperCase().includes("OANDA") && i + 1 < parts.length) {
      return parts[i + 1].toUpperCase();
    }
  }
  return stem;
}

function inferInterval(bars: Bar[]): number {
  if (bars.length < 2) return 0;
  const counts = new Map<number, number>();
  for (let i = 1; i < bars.length; i++) {
    const d = Math.round((bars[i].ms - bars[i - 1].ms) / 60000);
    counts.set(d, (counts.get(d) ?? 0) + 1);
  }
  let modal = 0;
  let best = -1;
  for (const [d, c] of counts) {
    if (c > best) {
      best = c;
      modal = d;
    }
  }
  return modal;
}

function extractOffset(iso: string): string {
  if (!iso) return "";
  if (iso.endsWith("Z")) return "+00:00";
  const tail = iso.slice(-6);
  if (tail.length === 6 && (tail[0] === "+" || tail[0] === "-") && tail[3] === ":") {
    return tail;
  }
  return "";
}

File: backend/app/data/loader.py
"""CSV loading, validation, and daylight-savings-correct timezone handling.

OANDA / TradingView exports use ISO-8601 timestamps that carry the chart's local
UTC offset (e.g. ``2026-05-19T15:00:00+02:00``). The offset reflects the export
zone, *not* the session zone, and it changes across DST boundaries. To keep
session-time logic (e.g. NY 09:30 open) correct, we always parse to a true UTC
instant first, then convert to the session zone via the tz database, which knows
each zone's DST transition dates.
"""
from __future__ import annotations

from dataclasses import dataclass, field
from io import BytesIO
from zoneinfo import ZoneInfo

import pandas as pd

REQUIRED_COLUMNS = ["time", "open", "high", "low", "close"]
# Canonical column names we expose downstream.
OHLC = ["open", "high", "low", "close"]


@dataclass
class Dataset:
    """A loaded, cleaned price series plus metadata.

    The frame is indexed by a tz-aware UTC ``DatetimeIndex`` and always contains
    ``open/high/low/close`` (float) and ``volume`` (float, 0 if absent).
    """

    df: pd.DataFrame
    instrument: str
    interval_minutes: int
    source_offset: str
    rows: int = field(init=False)

    def __post_init__(self) -> None:
        self.rows = len(self.df)

    def localize(self, tz: str) -> pd.DataFrame:
        """Return a copy of the frame with the index converted to ``tz``.

        Conversion goes through UTC, so DST is applied correctly by tzdata.
        """
        out = self.df.copy()
        out.index = out.index.tz_convert(ZoneInfo(tz))
        return out


def _infer_instrument(filename: str | None) -> str:
    if not filename:
        return "UNKNOWN"
    stem = filename.rsplit("/", 1)[-1].rsplit(".", 1)[0]
    # TradingView exports look like "OANDA_XAUUSD_30" or "5f7f2016-OANDA_XAUUSD_30".
    parts = stem.split("_")
    for i, p in enumerate(parts):
        # The OANDA token may be hyphen-prefixed, e.g. "5f7f2016-OANDA".
        if "OANDA" in p.upper() and i + 1 < len(parts):
            return parts[i + 1].upper()
    return stem


def _infer_interval_minutes(index: pd.DatetimeIndex) -> int:
    if len(index) < 2:
        return 0
    deltas = index.to_series().diff().dropna()
    if deltas.empty:
        return 0
    modal = deltas.mode()
    seconds = (modal.iloc[0] if not modal.empty else deltas.median()).total_seconds()
    return int(round(seconds / 60))


def load_csv(content: bytes, filename: str | None = None) -> Dataset:
    """Parse raw CSV bytes into a cleaned :class:`Dataset`.

    Raises ``ValueError`` on missing columns or unparseable timestamps.
    """
    raw = pd.read_csv(BytesIO(content))
    # Normalise column names: case-insensitive match to our schema.
    rename = {}
    lower_map = {c.lower(): c for c in raw.columns}
    for canonical in REQUIRED_COLUMNS + ["volume"]:
        if canonical in lower_map:
            rename[lower_map[canonical]] = canonical
    raw = raw.rename(columns=rename)

    missing = [c for c in REQUIRED_COLUMNS if c not in raw.columns]
    if missing:
        raise ValueError(
            f"CSV is missing required column(s): {', '.join(missing)}. "
            f"Found columns: {', '.join(raw.columns)}"
        )

    # Capture the source offset (for display) before normalising to UTC.
    sample_time = str(raw["time"].iloc[0]) if len(raw) else ""
    source_offset = _extract_offset(sample_time)

    # Parse to a true UTC instant regardless of the per-row offset.
    times = pd.to_datetime(raw["time"], utc=True, errors="coerce")
    if times.isna().any():
        bad = raw.loc[times.isna(), "time"].head(3).tolist()
        raise ValueError(f"Could not parse timestamp(s): {bad}")

    for col in OHLC:
        raw[col] = pd.to_numeric(raw[col], errors="coerce")
    raw["volume"] = (
        pd.to_numeric(raw["volume"], errors="coerce") if "volume" in raw.columns else 0.0
    )

    df = raw[OHLC + ["volume"]].copy()
    df.index = pd.DatetimeIndex(times)
    df = df[~df.index.duplicated(keep="last")].sort_index()
    df = df.dropna(subset=OHLC)

    if df.empty:
        raise ValueError("No valid rows remained after parsing.")

    return Dataset(
        df=df,
        instrument=_infer_instrument(filename),
        interval_minutes=_infer_interval_minutes(df.index),
        source_offset=source_offset,
    )


def _extract_offset(iso_time: str) -> str:
    """Pull the trailing ``+HH:MM`` / ``-HH:MM`` / ``Z`` offset for display."""
    if not iso_time:
        return ""
    if iso_time.endswith("Z"):
        return "+00:00"
    tail = iso_time[-6:]
    if len(tail) == 6 and tail[0] in "+-" and tail[3] == ":":
        return tail
    return ""
